src/InjectCoordinatesIntrinsics.cpp

```diff
@@ -0,0 +1,156 @@
+using std::vector;
+
+class InjectCoordinatesIntrinsics : public IRMutator {
+public:
+    InjectCoordinatesIntrinsics() : inside_kernel_loop(false) {}
+    Scope<int> scope;
+    bool inside_kernel_loop;
+
+private:
+    using IRMutator::visit;
+
+    void visit(const Provide *provide) {
+        if (!inside_kernel_loop) {
+            IRMutator::visit(provide);
+            return;
+        }
+
+        internal_assert(provide->values.size() == 1)
+            << "Coordinate currently only supports single-valued stores.\n";
+        user_assert(provide->args.size() == 3)
+            << "Coordinate stores require three coordinates.\n";
+
+        // Create coordinate_store("name", name.buffer, x, y, c, value)
+        // intrinsic.
+        Expr value_arg = mutate(provide->values[0]);
+        vector<Expr> args = {
+            provide->name,
+            Variable::make(Handle(), provide->name + ".buffer"),
+            provide->args[0],
+            provide->args[1],
+            provide->args[2],
+            value_arg};
+
+        stmt = Evaluate::make(Call::make(value_arg.type(),
+                                         Call::coordinates_store,
+                                         args,
+                                         Call::Intrinsic));
+    }
+
+    void visit(const Call *call) {
+        if (!inside_kernel_loop || call->call_type == Call::Intrinsic ||
+            call->call_type == Call::Extern) {
+            IRMutator::visit(call);
+            return;
+        }
+
+        string name = call->name;
+        if (call->call_type == Call::Halide && call->func.outputs() > 1) {
+            name = name + '.' + int_to_string(call->value_index);
+        }
+
+        vector<Expr> call_args = call->args;
+        // Check to see if we are reading from a one or two dimension function
+        // and pad to three dimensions.
+        while (call_args.size() < 3) {
+            call_args.push_back(IntImm::make(0));
+        }
+
+        // Create coordinates_load("name", "name[.n]", name.buffer, x, y, c)
+        // intrinsic call.
+        // We need to pass "name[.n]" because if we need to add normalization
+        // then we will this name as a prefix for "name[.n].extent" variable.
+        vector<Expr> args(6);
+        args[0] = call->name;
+        args[1] = name;
+        args[2] = Variable::make(Handle(), call->name + ".buffer");
+        for (size_t i = 0; i < call_args.size(); i++) {
+            string d = int_to_string(i);
+            string min_name = name + ".min." + d;
+            string min_name_constrained = min_name + ".constrained";
+            if (scope.contains(min_name_constrained)) {
+                min_name = min_name_constrained;
+            }
+
+            Expr min = Variable::make(Int(32), min_name);
+
+            // Remind users to explicitly specify the 'min' values of
+            // ImageParams accessed by coordinate-based filters.
+            if (i == 2 && call->param.defined()) {
```

Comment:
Yes, thanks. That was wrong call to move this from InjectOpenGL to InjectCoordinates.
However now I need to pass min name for that 3rd dimension to InjectOpenGL so it can preserve original OpenGL behavior and produce the warning.
I can replicate min-name creation in InjectOpenGL or I can stuff those names into intrinsic, but perhaps creating an IR node for this would be better?
Having IR node, where name creation can be encapsulated will also help with problem of passing "name[.n]" below.